

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>1_extractPCS module &mdash; pcsindels 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=d45e8c67"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2_computeWindows module" href="2_computeWindows.html" />
    <link rel="prev" title="Introduction" href="intro.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pcsindels
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">1_extractPCS module</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pre-requisites">Pre-requisites</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#a-chain-files-from-ucsc">a) Chain files from UCSC</a></li>
<li class="toctree-l3"><a class="reference internal" href="#b-fasta-files-from-ucsc">b) FASTA files from UCSC</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cluster-resources">Cluster resources</a></li>
<li class="toctree-l2"><a class="reference internal" href="#time-memory-disk-space">Time, Memory &amp; Disk space</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#time-per-run-details">Time per Run: Details</a></li>
<li class="toctree-l3"><a class="reference internal" href="#storage-per-run-details">Storage per Run: Details</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#genome-browser">Genome Browser</a></li>
<li class="toctree-l2"><a class="reference internal" href="#function-details">Function details</a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.PcsSingleCheck"><code class="docutils literal notranslate"><span class="pre">PcsSingleCheck</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#extractPCS.PcsSingleCheck.checkPCS"><code class="docutils literal notranslate"><span class="pre">PcsSingleCheck.checkPCS()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#extractPCS.PcsSingleCheck.print"><code class="docutils literal notranslate"><span class="pre">PcsSingleCheck.print()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#extractPCS.PcsSingleCheck.reset"><code class="docutils literal notranslate"><span class="pre">PcsSingleCheck.reset()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.checkPCSrevQ"><code class="docutils literal notranslate"><span class="pre">checkPCSrevQ()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.checkPCSrevT"><code class="docutils literal notranslate"><span class="pre">checkPCSrevT()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.checkPCSsize"><code class="docutils literal notranslate"><span class="pre">checkPCSsize()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.checkSpecies"><code class="docutils literal notranslate"><span class="pre">checkSpecies()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.computePCScoords"><code class="docutils literal notranslate"><span class="pre">computePCScoords()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.computePos"><code class="docutils literal notranslate"><span class="pre">computePos()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.initializePCSchecks"><code class="docutils literal notranslate"><span class="pre">initializePCSchecks()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.isCompatible"><code class="docutils literal notranslate"><span class="pre">isCompatible()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.performPCSchecks"><code class="docutils literal notranslate"><span class="pre">performPCSchecks()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.readChains"><code class="docutils literal notranslate"><span class="pre">readChains()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.readDNA"><code class="docutils literal notranslate"><span class="pre">readDNA()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#extractPCS.reverseComplement"><code class="docutils literal notranslate"><span class="pre">reverseComplement()</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="2_computeWindows.html">2_computeWindows module</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_setupEvolTimes.html">3_setupEvolTimes module</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_estimateEvolTimes.html">4_estimateEvolTimes module</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_sampleEvolTimes.html">5_sampleEvolTimes module</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_1_plotFig_PcsDistribComp.html">6_1_plotFig_PcsDistribComp module</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_2_plotFig_EvolTimesComp.html">6_2_plotFig_EvolTimesComp module</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_3_plotFig_MutRatesComp.html">6_3_plotFig_MutRatesComp module</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_4_plotFig_FuncClasses.html">6_4_plotFig_FuncClasses module</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pcsindels</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">1_extractPCS module</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/1_extractPCS.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-1_extractPCS">
<span id="extractpcs-module"></span><h1>1_extractPCS module<a class="headerlink" href="#module-1_extractPCS" title="Link to this heading"></a></h1>
<p>Reads a chain file (extension ‘.all.chain’) 
downloaded from the UCSC website 
and outputs one pickle file per chromosome with 
all PCSs of a pair of species.</p>
<p>To make sure that PCSs are being properly extracted,
this script also needs the FASTA files of both species, 
which can be downloaded in the UCSC website.</p>
<ul>
<li><p><strong>Use</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python3</span> <span class="mi">1</span><span class="n">_extractPCS</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">sp_ucsc_name</span> <span class="p">[</span><span class="n">UCSC</span> <span class="n">name</span><span class="p">]</span>
</pre></div>
</div>
</li>
<li><p><strong>Example of Usage (human (reference genome) and mouse)</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python3</span> <span class="o">~/</span><span class="n">code</span><span class="o">/</span><span class="mi">1</span><span class="n">_extractPCS</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">sp_ucsc_name</span> <span class="n">mm39</span>
</pre></div>
</div>
</li>
<li><p><strong>Input Parameter (mandatory)</strong>:</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">-sp_ucsc_name<span class="colon">:</span></dt>
<dd class="field-odd"><p>UCSC name of the species that is being aligned with the reference species (e.g. <em>mm39</em> for mouse).</p>
</dd>
</dl>
<ul class="simple">
<li><p><strong>Other Parameters taken from</strong> <code class="docutils literal notranslate"><span class="pre">dataset.py</span></code>:</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">refsp_ucscName<span class="colon">:</span></dt>
<dd class="field-odd"><p>UCSC name of the reference species that is being aligned (e.g. <em>hg38</em> for human).</p>
</dd>
<dt class="field-even">chain_file<span class="colon">:</span></dt>
<dd class="field-even"><p>file with extension <code class="docutils literal notranslate"><span class="pre">.all.chain</span></code> downloaded from UCSC (e.g.: hg38.papAnu4.all.chain.gz)</p>
</dd>
<dt class="field-odd">refsp_fastadir<span class="colon">:</span></dt>
<dd class="field-odd"><p>directory with FASTA files following the format <code class="docutils literal notranslate"><span class="pre">genomeName.chrName.bXXXeXXX.fa</span></code> (e.g. mm39.chrY.b51040000e51060000.fa).
FASTA files can be downloaded in the UCSC website, and then split into smaller files for more efficient reading.</p>
</dd>
<dt class="field-even">othsp_fastadir<span class="colon">:</span></dt>
<dd class="field-even"><p>directory with FASTA files following the format <code class="docutils literal notranslate"><span class="pre">genomeName.chrName.bXXXeXXX.fa</span></code> (e.g. mm39.chrY.b51040000e51060000.fa).
FASTA files can be downloaded in the UCSC website, and then split into smaller files for more efficient reading.</p>
</dd>
<dt class="field-odd">pcs_dir<span class="colon">:</span></dt>
<dd class="field-odd"><p>Directory where files with PCSs are going to be saved.</p>
</dd>
<dt class="field-even">pcs_minsize<span class="colon">:</span></dt>
<dd class="field-even"><p>minimum size (nb. of base pairs) for the PCS to be considered.</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Make sure that the required parameters described above are correctly defined in the file <code class="docutils literal notranslate"><span class="pre">utils/dataset.py</span></code>.</p>
</div>
<ul class="simple">
<li><dl class="simple">
<dt><strong>Output</strong>: </dt><dd><p>A separate <code class="docutils literal notranslate"><span class="pre">.pickle</span></code> file for each chromosome, each including all PCSs 
positioned on the corresponding chromosome of the reference genome. 
The <code class="docutils literal notranslate"><span class="pre">pickle</span></code> file saves a list of <code class="docutils literal notranslate"><span class="pre">Pcs</span></code> named tuples, where each tuple 
has the size of a PCS and its coordinates in both genomes.</p>
</dd>
</dl>
</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>Required Directory Structure for FASTA files</strong>: a single FASTA file 
from UCSC (e.g. <code class="docutils literal notranslate"><span class="pre">mm39.fa</span></code>) must be split in many files, 
following the pattern <code class="docutils literal notranslate"><span class="pre">genomeName.chrName.bXXXeXXX.fa</span></code> (see Pre-Requisites below).</p>
</div>
<section id="pre-requisites">
<h2>Pre-requisites<a class="headerlink" href="#pre-requisites" title="Link to this heading"></a></h2>
<blockquote>
<div><p>Before using this script, make sure all the required files were downloaded:</p>
</div></blockquote>
<section id="a-chain-files-from-ucsc">
<h3>a) Chain files from UCSC<a class="headerlink" href="#a-chain-files-from-ucsc" title="Link to this heading"></a></h3>
<p>1. Download the “chain” file for the analyzed pairwise species 
(chained lastz alignments) in the UCSC website. 
For example, for human and mouse (<em>hg38</em> and <em>mm39</em>), go to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">hgdownload</span><span class="o">.</span><span class="n">soe</span><span class="o">.</span><span class="n">ucsc</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">goldenPath</span><span class="o">/</span><span class="n">mm39</span><span class="o">/</span><span class="n">vsHg38</span><span class="o">/</span>
</pre></div>
</div>
<p>and download the file <code class="docutils literal notranslate"><span class="pre">mm39.hg38.all.chain.gz</span></code>.</p>
<ol class="arabic" start="2">
<li><p>Unzip the download file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gzip</span> <span class="o">-</span><span class="n">d</span> <span class="n">mm39</span><span class="o">.</span><span class="n">hg38</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">chain</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</li>
</ol>
<p>3. The full directory of the unziped chain file (e.g. <code class="docutils literal notranslate"><span class="pre">mm39.hg38.all.chain</span></code>) 
will be given to this script through the parameter <code class="docutils literal notranslate"><span class="pre">-chain_file</span></code>.</p>
</section>
<section id="b-fasta-files-from-ucsc">
<h3>b) FASTA files from UCSC<a class="headerlink" href="#b-fasta-files-from-ucsc" title="Link to this heading"></a></h3>
<p>4. Download the FASTA files for the analyzed pairwise species in the UCSC website (e.g. <code class="docutils literal notranslate"><span class="pre">mm39.fa.gz</span></code>).
The FASTA files are needed to validate the PCS coordinates obtained from the chain file.</p>
<p>5. Run the script <code class="docutils literal notranslate"><span class="pre">breakFasta.py</span></code> using the unziped FASTA file (e.g. <code class="docutils literal notranslate"><span class="pre">mm39.fa</span></code>). 
This script will <em>break</em> the unziped FASTA file into smaller files, following the pattern:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genomeName</span><span class="o">.</span><span class="n">chrName</span><span class="o">.</span><span class="n">bXXXeXXX</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
<ol class="arabic" start="6">
<li><p>These files must be saved in a directory structure organized as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genomeName</span><span class="o">/</span><span class="nb">chr</span><span class="o">/</span><span class="n">chrName</span><span class="o">/</span><span class="n">genomeName</span><span class="o">.</span><span class="n">chrName</span><span class="o">.</span><span class="n">bXXXeXXX</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
</li>
</ol>
</section>
</section>
<section id="cluster-resources">
<h2>Cluster resources<a class="headerlink" href="#cluster-resources" title="Link to this heading"></a></h2>
<p>In case you want to run this Python script stand-alone in a cluster that uses Slurm to manage jobs:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">srun</span> <span class="o">-</span><span class="n">p</span> <span class="n">compute</span> <span class="o">-</span><span class="n">t</span> <span class="mi">2</span><span class="p">:</span><span class="mi">00</span><span class="p">:</span><span class="mi">00</span> <span class="o">--</span><span class="n">mem</span> <span class="mi">20</span><span class="n">G</span> <span class="o">--</span><span class="n">nodes</span><span class="o">=</span><span class="mi">1</span> <span class="o">--</span><span class="n">ntasks</span><span class="o">=</span><span class="mi">1</span> <span class="o">--</span><span class="n">cpus</span><span class="o">-</span><span class="n">per</span><span class="o">-</span><span class="n">task</span><span class="o">=</span><span class="mi">1</span> <span class="o">--</span><span class="n">pty</span> <span class="n">bash</span>
</pre></div>
</div>
<p>Otherwise you can use the script <code class="docutils literal notranslate"><span class="pre">../cluster/1_extractPCS_runAll.py</span></code>
to run this script for all 40 species used in this study.</p>
</section>
<section id="time-memory-disk-space">
<h2>Time, Memory &amp; Disk space<a class="headerlink" href="#time-memory-disk-space" title="Link to this heading"></a></h2>
<p>For reference, here we include an <strong>upper limit</strong> on runtime, memory usage, and disk 
space required for running this script on each pairwise alignment of the 40 
vertebrate dataset examined in our study.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>UCSC name</p></th>
<th class="head"><p>Time</p></th>
<th class="head"><p>Memory</p></th>
<th class="head"><p>Disk</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>panPan3</p></td>
<td><p>01:01:19</p></td>
<td><p>5GB</p></td>
<td><p>0.95GB</p></td>
</tr>
<tr class="row-odd"><td><p>panTro6</p></td>
<td><p>00:59:09</p></td>
<td><p>6GB</p></td>
<td><p>0.98GB</p></td>
</tr>
<tr class="row-even"><td><p>gorGor6</p></td>
<td><p>00:50:02</p></td>
<td><p>5GB</p></td>
<td><p>1.14GB</p></td>
</tr>
<tr class="row-odd"><td><p>ponAbe3</p></td>
<td><p>01:20:13</p></td>
<td><p>5GB</p></td>
<td><p>1.98GB</p></td>
</tr>
<tr class="row-even"><td><p>papAnu4</p></td>
<td><p>06:27:13</p></td>
<td><p>38GB</p></td>
<td><p>2.76GB</p></td>
</tr>
<tr class="row-odd"><td><p>macFas5</p></td>
<td><p>02:42:13</p></td>
<td><p>20GB</p></td>
<td><p>2.71GB</p></td>
</tr>
<tr class="row-even"><td><p>rhiRox1</p></td>
<td><p>03:29:32</p></td>
<td><p>23GB</p></td>
<td><p>2.98GB</p></td>
</tr>
<tr class="row-odd"><td><p>chlSab2</p></td>
<td><p>00:59:52</p></td>
<td><p>6GB</p></td>
<td><p>2.99GB</p></td>
</tr>
<tr class="row-even"><td><p>nasLar1</p></td>
<td><p>04:26:24</p></td>
<td><p>15GB</p></td>
<td><p>2.25GB</p></td>
</tr>
<tr class="row-odd"><td><p>rheMac10</p></td>
<td><p>01:01:55</p></td>
<td><p>6GB</p></td>
<td><p>2.80GB</p></td>
</tr>
<tr class="row-even"><td><p>calJac4</p></td>
<td><p>02:50:19</p></td>
<td><p>36GB</p></td>
<td><p>3.14GB</p></td>
</tr>
<tr class="row-odd"><td><p>tarSyr2</p></td>
<td><p>04:57:09</p></td>
<td><p>42GB</p></td>
<td><p>2.62GB</p></td>
</tr>
<tr class="row-even"><td><p>micMur2</p></td>
<td><p>02:24:55</p></td>
<td><p>17GB</p></td>
<td><p>2.25GB</p></td>
</tr>
<tr class="row-odd"><td><p>galVar1</p></td>
<td><p>09:48:31</p></td>
<td><p>40GB</p></td>
<td><p>2.56GB</p></td>
</tr>
<tr class="row-even"><td><p>mm39</p></td>
<td><p>00:58:45</p></td>
<td><p>9GB</p></td>
<td><p>1.10GB</p></td>
</tr>
<tr class="row-odd"><td><p>oryCun2</p></td>
<td><p>01:23:21</p></td>
<td><p>19GB</p></td>
<td><p>1.67GB</p></td>
</tr>
<tr class="row-even"><td><p>rn7</p></td>
<td><p>01:27:01</p></td>
<td><p>24GB</p></td>
<td><p>1.07GB</p></td>
</tr>
<tr class="row-odd"><td><p>vicPac2</p></td>
<td><p>01:23:54</p></td>
<td><p>17GB</p></td>
<td><p>2.19GB</p></td>
</tr>
<tr class="row-even"><td><p>bisBis1</p></td>
<td><p>02:50:09</p></td>
<td><p>24GB</p></td>
<td><p>1.86GB</p></td>
</tr>
<tr class="row-odd"><td><p>felCat9</p></td>
<td><p>01:57:19</p></td>
<td><p>28GB</p></td>
<td><p>2.19GB</p></td>
</tr>
<tr class="row-even"><td><p>manPen1</p></td>
<td><p>03:18:57</p></td>
<td><p>18GB</p></td>
<td><p>1.99GB</p></td>
</tr>
<tr class="row-odd"><td><p>bosTau9</p></td>
<td><p>01:10:49</p></td>
<td><p>15GB</p></td>
<td><p>1.59GB</p></td>
</tr>
<tr class="row-even"><td><p>canFam6</p></td>
<td><p>01:00:37</p></td>
<td><p>18GB</p></td>
<td><p>2.07GB</p></td>
</tr>
<tr class="row-odd"><td><p>musFur1</p></td>
<td><p>01:00:55</p></td>
<td><p>17GB</p></td>
<td><p>2.21GB</p></td>
</tr>
<tr class="row-even"><td><p>neoSch1</p></td>
<td><p>01:20:42</p></td>
<td><p>19GB</p></td>
<td><p>2.46GB</p></td>
</tr>
<tr class="row-odd"><td><p>equCab3</p></td>
<td><p>01:30:38</p></td>
<td><p>17GB</p></td>
<td><p>2.60GB</p></td>
</tr>
<tr class="row-even"><td><p>myoLuc2</p></td>
<td><p>01:46:55</p></td>
<td><p>20GB</p></td>
<td><p>1.60GB</p></td>
</tr>
<tr class="row-odd"><td><p>susScr11</p></td>
<td><p>01:20:55</p></td>
<td><p>18GB</p></td>
<td><p>1.93GB</p></td>
</tr>
<tr class="row-even"><td><p>enhLutNer1</p></td>
<td><p>01:27:55</p></td>
<td><p>18GB</p></td>
<td><p>2.26GB</p></td>
</tr>
<tr class="row-odd"><td><p>triMan1</p></td>
<td><p>01:20:32</p></td>
<td><p>20GB</p></td>
<td><p>2.16GB</p></td>
</tr>
<tr class="row-even"><td><p>macEug2</p></td>
<td><p>05:58:51</p></td>
<td><p>18GB</p></td>
<td><p>0.22GB</p></td>
</tr>
<tr class="row-odd"><td><p>ornAna2</p></td>
<td><p>01:20:04</p></td>
<td><p>13GB</p></td>
<td><p>0.17GB</p></td>
</tr>
<tr class="row-even"><td><p>aptMan1</p></td>
<td><p>00:45:15</p></td>
<td><p>3GB</p></td>
<td><p>0.12GB</p></td>
</tr>
<tr class="row-odd"><td><p>galGal6</p></td>
<td><p>00:45:15</p></td>
<td><p>3GB</p></td>
<td><p>0.09GB</p></td>
</tr>
<tr class="row-even"><td><p>thaSir1</p></td>
<td><p>01:30:08</p></td>
<td><p>5GB</p></td>
<td><p>0.07GB</p></td>
</tr>
<tr class="row-odd"><td><p>aquChr2</p></td>
<td><p>00:30:03</p></td>
<td><p>3GB</p></td>
<td><p>0.12GB</p></td>
</tr>
<tr class="row-even"><td><p>melGal5</p></td>
<td><p>01:15:00</p></td>
<td><p>3GB</p></td>
<td><p>0.09GB</p></td>
</tr>
<tr class="row-odd"><td><p>xenLae2</p></td>
<td><p>01:15:46</p></td>
<td><p>6GB</p></td>
<td><p>0.06GB</p></td>
</tr>
<tr class="row-even"><td><p>xenTro10</p></td>
<td><p>02:40:30</p></td>
<td><p>40GB</p></td>
<td><p>0.06GB</p></td>
</tr>
<tr class="row-odd"><td><p>danRer11</p></td>
<td><p>04:08:26</p></td>
<td><p>14GB</p></td>
<td><p>0.04GB</p></td>
</tr>
</tbody>
</table>
<section id="time-per-run-details">
<h3>Time per Run: Details<a class="headerlink" href="#time-per-run-details" title="Link to this heading"></a></h3>
<p>Stats on time of a single run (human-mouse alignment, whole-genome): <strong>~65 minutes</strong>
Details on computational time are available in the log of the run.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Time (s)</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Reading chains</p></td>
<td><p>164.45332742</p></td>
</tr>
<tr class="row-odd"><td><p>Compatible chains</p></td>
<td><p>236.70236850</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr1</p></td>
<td><p>246.49562716</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr2</p></td>
<td><p>283.03650451</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr3</p></td>
<td><p>218.50386620</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr4</p></td>
<td><p>198.70389938</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr5</p></td>
<td><p>206.89367223</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr6</p></td>
<td><p>199.71864271</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr7</p></td>
<td><p>204.75429463</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr8</p></td>
<td><p>199.65774012</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr9</p></td>
<td><p>150.57288957</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr10</p></td>
<td><p>173.72448373</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr11</p></td>
<td><p>156.16801929</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr12</p></td>
<td><p>168.79578829</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr13</p></td>
<td><p>113.74493265</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr14</p></td>
<td><p>93.78934121</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr15</p></td>
<td><p>99.19000816</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr16</p></td>
<td><p>91.97542357</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr17</p></td>
<td><p>85.33002925</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr18</p></td>
<td><p>83.99309540</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr19</p></td>
<td><p>102.53493857</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr20</p></td>
<td><p>72.90905333</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chr21</p></td>
<td><p>37.42282581</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chr22</p></td>
<td><p>44.24664927</p></td>
</tr>
<tr class="row-even"><td><p>PCSs from chrX</p></td>
<td><p>223.61548066</p></td>
</tr>
<tr class="row-odd"><td><p>PCSs from chrY</p></td>
<td><p>21.73898244</p></td>
</tr>
<tr class="row-even"><td><p><strong>Total time</strong></p></td>
<td><p>3949.23640776</p></td>
</tr>
</tbody>
</table>
</section>
<section id="storage-per-run-details">
<h3>Storage per Run: Details<a class="headerlink" href="#storage-per-run-details" title="Link to this heading"></a></h3>
<p>Size of output files with PCSs for the <em>human</em> and <em>mouse</em> 
alignment (whole-genome): <strong>~1085 MB</strong>.</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Output file</p></th>
<th class="head"><p>Size</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>hg38.mm39.chr1.pcs.pickle</p></td>
<td><p>98M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr2.pcs.pickle</p></td>
<td><p>104M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr3.pcs.pickle</p></td>
<td><p>72M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr4.pcs.pickle</p></td>
<td><p>59M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr5.pcs.pickle</p></td>
<td><p>74M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr6.pcs.pickle</p></td>
<td><p>66M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr7.pcs.pickle</p></td>
<td><p>56M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr8.pcs.pickle</p></td>
<td><p>55M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr9.pcs.pickle</p></td>
<td><p>44M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr10.pcs.pickle</p></td>
<td><p>46M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr11.pcs.pickle</p></td>
<td><p>65M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr12.pcs.pickle</p></td>
<td><p>50M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr13.pcs.pickle</p></td>
<td><p>33M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr14.pcs.pickle</p></td>
<td><p>39M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr15.pcs.pickle</p></td>
<td><p>34M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr16.pcs.pickle</p></td>
<td><p>30M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr17.pcs.pickle</p></td>
<td><p>24M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr18.pcs.pickle</p></td>
<td><p>30M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr19.pcs.pickle</p></td>
<td><p>14M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr20.pcs.pickle</p></td>
<td><p>28M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chr21.pcs.pickle</p></td>
<td><p>12M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chr22.pcs.pickle</p></td>
<td><p>11M</p></td>
</tr>
<tr class="row-even"><td><p>hg38.mm39.chrX.pcs.pickle</p></td>
<td><p>41M</p></td>
</tr>
<tr class="row-odd"><td><p>hg38.mm39.chrY.pcs.pickle</p></td>
<td><p>387K</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="genome-browser">
<h2>Genome Browser<a class="headerlink" href="#genome-browser" title="Link to this heading"></a></h2>
<p>Here there are instructions in case you wish to check the corresponding 
DNA sequences of a PCS in UCSC’s Genome Browser. Take for example the 
following PCS found in the pairwise alignment between human (hg38) and bonobo (panPan3):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Pcs</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">tChrom</span><span class="o">=</span><span class="s1">&#39;chr10&#39;</span><span class="p">,</span> <span class="n">tStrand</span><span class="o">=</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="n">tPosBeg</span><span class="o">=</span><span class="mi">87682</span><span class="p">,</span> <span class="n">qChrom</span><span class="o">=</span><span class="s1">&#39;chr16&#39;</span><span class="p">,</span> <span class="n">qStrand</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">qPosBeg</span><span class="o">=</span><span class="mi">21852</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, <code class="docutils literal notranslate"><span class="pre">qChrom</span></code>, <code class="docutils literal notranslate"><span class="pre">qStrand</span></code>, and <code class="docutils literal notranslate"><span class="pre">qPosBeg</span></code> correspond to the chromosome, strand, 
and starting coordinate in the query genome. In our dataset, the query genome is 
always the human genome (hg38). Similarly, <code class="docutils literal notranslate"><span class="pre">tChrom</span></code>, <code class="docutils literal notranslate"><span class="pre">tStrand</span></code>, and <code class="docutils literal notranslate"><span class="pre">tPosBeg</span></code> 
refer to the same information in the target genome (panPan3 in this example).</p>
<p>To compute the coordinates, first add <code class="docutils literal notranslate"><span class="pre">+1</span></code> to <code class="docutils literal notranslate"><span class="pre">tPosBeg</span></code> and <code class="docutils literal notranslate"><span class="pre">qPosBeg</span></code>, as the 
coordinates in <em>Genome Browser</em> start at 1, whereas here they start at 0.</p>
<p>Next, by summing the size of the PCS (size=45) to 
the initial coordinates (<code class="docutils literal notranslate"><span class="pre">tPosBeg+1</span></code>  and <code class="docutils literal notranslate"><span class="pre">qPosBeg+1</span></code>), it is possible 
to retrieve the DNA sequence in Genome Browser for both genomes.</p>
<p>For this example, by accessing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38
</pre></div>
</div>
<p>and typing <code class="docutils literal notranslate"><span class="pre">chr16:21853-21897</span></code>, we obtain:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TAATGAGTGTACTGCTCCTAAACAAGGAGTATCTGCATTTATTTA</span>
</pre></div>
</div>
<p>If it is indicated that sequence should be read in the negative strand (qStrand=’-‘),
the reverse complement should be computed:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TAAATAAATGCAGATACTCCTTGTTTAGGAGCAGTACACTCATTA</span>
</pre></div>
</div>
<p>Similarly for the bonobo genome, by accessing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>https://genome.ucsc.edu/cgi-bin/hgTracks?db=panPan3
</pre></div>
</div>
<p>and typing <code class="docutils literal notranslate"><span class="pre">chr10:87683-87727</span></code>, we obtain:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TAAATAAATGCAGATACTCCTTGTTTAGGAGCAGTACACTCATTA</span>
</pre></div>
</div>
</section>
<section id="function-details">
<h2>Function details<a class="headerlink" href="#function-details" title="Link to this heading"></a></h2>
<p>Only relevant functions have been documented below. 
For more details on any function, check the comments in the souce code.</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="extractPCS.PcsSingleCheck">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">PcsSingleCheck</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">desc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cntUB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pcsSingleCheckFunc</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.PcsSingleCheck" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This class makes sure that PCSs are being correctly processed. 
For every new PCS, it determines whether the PCS is a good candidate 
to be checked into details. Good candidates are long PCSs,
PCSs in the reverse strand of one of the genomes, among others.
It keeps a history of how many PCSs have been checked so far.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>desc</strong> (<em>str</em>) – Brief description of which type of check is being performed.</p></li>
<li><p><strong>cntUB</strong> (<em>int</em>) – Maximum number of PCSs that will be checked in details.</p></li>
<li><p><strong>pcsSingleCheckFunc</strong> (<em>function</em>) – Function that receives a named tuple PCS 
and returns True (PCS should be checked) or False.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="extractPCS.PcsSingleCheck.checkPCS">
<span class="sig-name descname"><span class="pre">checkPCS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pcs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.PcsSingleCheck.checkPCS" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="extractPCS.PcsSingleCheck.print">
<span class="sig-name descname"><span class="pre">print</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.PcsSingleCheck.print" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="extractPCS.PcsSingleCheck.reset">
<span class="sig-name descname"><span class="pre">reset</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.PcsSingleCheck.reset" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.checkPCSrevQ">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">checkPCSrevQ</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pcs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.checkPCSrevQ" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.checkPCSrevT">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">checkPCSrevT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pcs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.checkPCSrevT" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.checkPCSsize">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">checkPCSsize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pcs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.checkPCSsize" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.checkSpecies">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">checkSpecies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ucscName</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.checkSpecies" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.computePCScoords">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">computePCScoords</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">begAbs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">posRel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isRevComp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sizePCS</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sizeMis</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.computePCScoords" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.computePos">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">computePos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.computePos" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.initializePCSchecks">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">initializePCSchecks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cntUB</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefixTarget</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tar'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefixQuery</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'qry'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.initializePCSchecks" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.isCompatible">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">isCompatible</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos_lst</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.isCompatible" title="Link to this definition"></a></dt>
<dd><p>Check if a sequence is compatible (non-intersecting) to a list of sequences.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.performPCSchecks">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">performPCSchecks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pcs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">checks</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.performPCSchecks" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.readChains">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">readChains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chainsFilename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.readChains" title="Link to this definition"></a></dt>
<dd><p>This function parses the chains from a chain file (check the UCSC 
website for more information on the format).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.readDNA">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">readDNA</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirFASTA</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrName</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">posBeg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">posEnd</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debug</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.readDNA" title="Link to this definition"></a></dt>
<dd><p>This function reads a DNA segment in a FASTA file.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="extractPCS.reverseComplement">
<span class="sig-prename descclassname"><span class="pre">1_extractPCS.</span></span><span class="sig-name descname"><span class="pre">reverseComplement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#extractPCS.reverseComplement" title="Link to this definition"></a></dt>
<dd><p>This function returns a reverse complement of a DNA.</p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="intro.html" class="btn btn-neutral float-left" title="Introduction" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="2_computeWindows.html" class="btn btn-neutral float-right" title="2_computeWindows module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Priscila Biller.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>